File: MyFriends/TableViewController.m
//
//  TableViewController.m
//  MyFriends
//
//  Created by Bobby Towers on 1/23/15.
//  Copyright (c) 2015 Bobby Towers. All rights reserved.
//

#import "TableViewController.h"

#import <Parse/Parse.h>

@interface TableViewController ()

@property (weak, nonatomic) IBOutlet UITextField *friendTextField;

@end

@implementation TableViewController
{
    
    NSMutableArray * items;
    
}


- (void)viewDidLoad {
    [super viewDidLoad];
    
    [PFUser currentUser].username = @"Bobby";
    
    [PFUser currentUser].password = @"password";
    
    [[PFUser currentUser] saveInBackground];
    
    items = [@[] mutableCopy];
    
    [NSTimer scheduledTimerWithTimeInterval:5 target:self selector:@selector(updateList) userInfo:nil repeats:YES];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
}

- (void)updateList {
    
    PFQuery * listQuery = [PFQuery queryWithClassName:@"Message"];
    
    [listQuery orderByDescending:@"createdAt"];
    
    //    [messageQuery whereKey:@"owner" equalTo:[PFUser currentUser]];
    
    [listQuery includeKey:@"owner"];
    
    [listQuery findObjectsInBackgroundWithBlock:^(NSArray *objects, NSError *error) {
        
        if (objects.count > 0) {
            
            items = [objects mutableCopy];
            
            //            NSLog(@"found %@", messages);
            
            [self.tableView reloadData];
            
        }
        
    }];
    
}

- (IBAction)addItem:(id)sender {
    
    // PFObject is a blank template for any class name or item name
    PFObject * newItem = [PFObject objectWithClassName:@"Item"];
    
    newItem[@"text"] = self.friendTextField.text;
    
    // this is a singleton
    // a singleton alloc and inits once
    newItem[@"owner"] = [PFUser currentUser];
    
    [newItem saveInBackground];
    
    self.friendTextField.text = @"";
    
    // adds item to items array and reloads the tableView
    //    [items addObject:newItem];
    
    [items insertObject:newItem atIndex:0];
    
    [self.tableView reloadData];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

//- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
//#warning Potentially incomplete method implementation.
//    // Return the number of sections.
//    return 0;
//}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
#warning Incomplete method implementation.
    // Return the number of rows in the section.
    return items.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"itemCell" forIndexPath:indexPath];
    
    PFObject * message = items[indexPath.row];
    PFUser * owner = message[@"owner"];
    
    cell.textLabel.text = owner.username;
    cell.detailTextLabel.text = message[@"text"];
    
    return cell;
}


/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/


// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        
        PFObject * item = items[indexPath.row];
        
        [item deleteInBackground];
        
        [items removeObjectAtIndex:indexPath.row];
        
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }
}


/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: FirebaseChat/FirebaseChat/TableViewController.swift
//
//  TableViewController.swift
//  FirebaseChat
//
//  Created by Bobby Towers on 3/12/15.
//  Copyright (c) 2015 Bobby Towers. All rights reserved.
//

import UIKit

let firebaseRef = Firebase(url:"https://jo-chat.firebaseio.com/")


class TableViewController: UITableViewController {

    
    @IBOutlet weak var messageField: UITextField!
    
    @IBAction func sendMessage(sender: AnyObject) {
        
        var dateRounded = Int(NSDate().timeIntervalSince1970)
        
        println(dateRounded)
        
        firebaseRef.childByAppendingPath("messages").childByAppendingPath("\(dateRounded)_Swagarino").setValue(["name":"Captain Crunch", "message":messageField.text])
        
        messageField.text = ""
    }
    
    var messages: [[String:AnyObject]] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Web sockets are like phone conversations
        // It is waiting for events to be passed, and it's an ongoing conversation
        // As opposed to an email-like conversation
        
        // Creating the observer in the viewDidLoad 
        // Almost exactly like a notification center/observer
        // Setting it up, runs itself when a new message comes in
        
        firebaseRef.childByAppendingPath("messages").observeEventType(.ChildAdded, withBlock: {
            snapshot in
            
            println("\(snapshot.key) -> \(snapshot.value)")
            
            self.messages.insert(snapshot.value as [String:AnyObject], atIndex: 0)
//            self.messages.append(snapshot.value as [String:AnyObject])
            
            self.tableView.reloadData()
        })
        
//        firebaseRef.setValue("Swagarino Cappuccino")
        
        
        
        
/*
        let block: (snapshot: FDataSnapshot!) -> Void = {
            snapshot in
            
            println("\(snapshot.key) -> \(snapshot.value)")
            
            self.messages.insert(snapshot.value as [String:AnyObject], atIndex: 0)
            
            self.tableView.reloadData()
            
        }
        
        
        firebaseRef.childByAppendingPath("messages").observeEventType(.ChildChanged, withBlock: block)
        firebaseRef.childByAppendingPath("messages").observeEventType(.ChildAdded, withBlock: block)
        firebaseRef.childByAppendingPath("messages").observeEventType(.ChildRemoved, withBlock: block)
*/
        
        
        
        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false

        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source

//    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
//        // #warning Potentially incomplete method implementation.
//        // Return the number of sections.
//        return 0
//    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete method implementation.
        // Return the number of rows in the section.
        return messages.count
    }

    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCellWithIdentifier("messageCell", forIndexPath: indexPath) as UITableViewCell

        let message = messages[indexPath.row]
        
        println(message)
        
        cell.textLabel?.text = message["name"] as? String
        cell.detailTextLabel?.text = message["message"] as? String
        
        return cell
    }


    /*
    // Override to support conditional editing of the table view.
    override func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Return NO if you do not want the specified item to be editable.
        return true
    }
    */

    /*
    // Override to support editing the table view.
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        if editingStyle == .Delete {
            // Delete the row from the data source
            tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Fade)
        } else if editingStyle == .Insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
        }    
    }
    */

    /*
    // Override to support rearranging the table view.
    override func tableView(tableView: UITableView, moveRowAtIndexPath fromIndexPath: NSIndexPath, toIndexPath: NSIndexPath) {

    }
    */

    /*
    // Override to support conditional rearranging of the table view.
    override func tableView(tableView: UITableView, canMoveRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Return NO if you do not want the item to be re-orderable.
        return true
    }
    */

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using [segue destinationViewController].
        // Pass the selected object to the new view controller.
    }
    */

}
